### PR TITLE
added option to print url, fixes #344

### LINKED ISSUE #344: Add option to print final url to standard output
It would be helpful to have an option to let mozdownload print the final url to stdout. It could be used with pipes to pass into another application or stored to disk.

### PR DESCRIPTION
**Pull request by Prashant-Surya**
Add option to print final url to standard output #344

### CODE REVIEW

File: mozdownload/cli.py [hunk 1]
```diff
@@ -119,6 +119,10 @@ def parse_arguments(argv):
                         help='Version of the application to be downloaded for release '
                              'and candidate builds (special values: %s)' % ', '.join(
                                 scraper.RELEASE_AND_CANDIDATE_LATEST_VERSIONS.keys()))
+    parser.add_argument('--print-url',
+                        action='store_true',
+                        help="Prints final url.")
```
Comment (whimboo) [inline mozdownload/cli.py]: I would suggest to use 'Print final URL instead of downloading the file'. Please also add `dest` so we stay in sync with all the other options.

File: mozdownload/cli.py [hunk 2]
```diff
@@ -142,7 +146,6 @@ def parse_arguments(argv):
 def cli(argv=None):
     """CLI entry point for mozdownload."""
     kwargs = parse_arguments(argv or sys.argv[1:])
-
```
Comment (whimboo) [inline mozdownload/cli.py]: nit: Unnecessary change. Please revert this empty line.

File: mozdownload/cli.py [hunk 3]
```diff
@@ -163,7 +166,7 @@ def cli(argv=None):
             scraper_type = 'direct'
 
         build = factory.FactoryScraper(scraper_type, **kwargs)
-        build.download()
+        build.download(kwargs['print_url'])
```
Comment (whimboo) [inline mozdownload/cli.py]: We should not call download if we only want to print the URL. Add the if condition here instead.

File: mozdownload/scraper.py [hunk 4]
```diff
@@ -266,12 +266,14 @@ def total_seconds(td):
         self.logger.info('Downloading from: %s' % self.url)
         self.logger.info('Saving as: %s' % self.filename)
 
+        if print_url:
+            print self.url
```
Comment (whimboo) [inline mozdownload/scraper.py]: Please see the line 266 above. We already print the final URL here. The issue itself is that we want to print the URL without downloading the build. Sorry if that was not that clear in the description.

File: mozdownload/scraper.py [hunk 5]
```diff
@@ -266,12 +266,14 @@ def total_seconds(td):
...
+
         tmp_file = self.filename + ".part"
 
         def _download():
             try:
                 start_time = datetime.now()
-
```
Comment (whimboo) [inline mozdownload/scraper.py]: nit: Same comment as above for the removal of the empty line.

**Comment by Prashant-Surya** [comment]
I'll make those changes and update the pull request.

**Comment by whimboo** [comment]
Sorry that I missed your latest updates for this PR. @Prashant-Surya will you be able to finish it off? I will add some inline comments now.

File: mozdownload/cli.py [hunk 6]
```diff
@@ -119,6 +119,10 @@ def parse_arguments(argv):
                         help='Version of the application to be downloaded for release '
                              'and candidate builds (special values: %s)' % ', '.join(
                                 scraper.RELEASE_AND_CANDIDATE_LATEST_VERSIONS.keys()))
+    parser.add_argument('--print-url',
```
Comment (whimboo) [inline mozdownload/cli.py]: nit: All the arguments are sorted alphabetically. Can you please move this block up? Thanks.

**Comment by Prashant-Surya** [comment]
@whimboo sure. I'll make changes

File: mozdownload/cli.py [hunk 7]
```diff
@@ -163,7 +167,10 @@ def cli(argv=None):
             scraper_type = 'direct'
 
         build = factory.FactoryScraper(scraper_type, **kwargs)
-        build.download()
+        if kwargs.get('print_url'):
+            print build.url
```
Comment (whimboo) [inline mozdownload/cli.py]: Instead of `print` you should use a call to `logger.info()` so that it get printed correctly.

File: tests/cli/test_cli_print_url.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    def test_print_url_argument(self):
+        try:
+            args = ['mozdownload',
+                '--url=https://raw.github.com/mozilla/mozdownload/master/README.md',
+                '--print-url']
+            output = subprocess.check_output(args)
+        except subprocess.CalledProcessError as e:
+            output = e.output
+
+        url = r'https://raw.github.com/mozilla/mozdownload/master/README.md'
```
Comment (whimboo) [inline tests/cli/test_cli_print_url.py]: Please define `url` above so you do not have to hard-code it in line 17.

**Comment by whimboo** [comment]
Thanks a lot, and sorry again for the delay. If we can get this PR landed soon, it will even make the next release of mozdownload 1.22.

**Comment by Prashant-Surya** [comment]
I'll try to finish it asap

**Comment by Prashant-Surya** [comment]
@whimboo Hi I tried changing the test case. As I'm printing url using logging.info should I use mock for checking URL??

**Comment by whimboo** [comment]
It basically shouldn't matter for the test how the content gets printed. When using `logging.info` it also appears in stdout, and the regex match should find it.

**Comment by Prashant-Surya** [comment]
@whimboo I updated the code can u check?

File: mozdownload/cli.py [hunk 9]
```diff
@@ -146,7 +150,6 @@ def cli(argv=None):
     log_level = kwargs.pop('log_level')
     logging.basicConfig(format='%(levelname)s | %(message)s', level=log_level)
     logger = logging.getLogger(__name__)
-
```
Comment (whimboo) [inline mozdownload/cli.py]: nit: Please leave this line untouched. Thanks.

File: tests/cli/test_cli_print_url.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    def test_print_url_argument(self):
+        try:
+            args = ['mozdownload',
+                '--url=https://raw.github.com/mozilla/mozdownload/master/README.md',
+                '--print-url']
+            output = subprocess.check_output(args, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as e:
+            output = e.output
+
+        url = r'https://raw.github.com/mozilla/mozdownload/master/README.md'
```
Comment (whimboo) [inline tests/cli/test_cli_print_url.py]: Something which I already mentioned before is that you should define this URL only once. You can pass it into the call to `mozdownload` and also use it for the assertion check later.

**Comment by Prashant-Surya** [comment]
@whimboo Fixed those changes. Sorry. Forgot to add my changes in test file.

File: tests/cli/test_cli_print_url.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import unittest
+
+
+class TestPrintUrlArgument(unittest.TestCase):
+    """Test CLI for print-url argument."""
+
+    def test_print_url_argument(self):
+        url = r'https://raw.github.com/mozilla/mozdownload/master/README.md'
+        try:
+            url_arg = '--url={url}'.format(url=url)
```
Comment (whimboo) [inline tests/cli/test_cli_print_url.py]: Ok, one last thing then we can land it. Please bear with me. :)

There is no need for this extra variable. Just move this line into the list below. Thanks.

**Comment by whimboo** [comment]
Sorry for the late reply but Github didn't inform me about the recent push. @Prashant-Surya best is if you always add an extra comment once you pushed a change, and the PR is ready for the next review.

This patch looks fine now, and I'm going to get it merged. Let me know if you have interests to work on something else, or check our open issues. Thanks!

**Comment by whimboo** [comment]
@Prashant-Surya surely! What do you think about issue #247? It's not that complicated and would be a good next step for you.

**Comment by Prashant-Surya** [comment]
@whimboo But someone already worked on it I think. If it's still is open I'll take it up

**Comment by whimboo** [comment]
He abandoned the work. So you can definitely pick it up.

**Comment by Prashant-Surya** [comment]
Okay. I'll take it up.
